ic-asset/src/asset_config.rs

```diff
@@ -0,0 +1,626 @@
+    Ok(())
+}
+
+impl AssetConfigRule {
+    fn applies(&self, canonical_path: &Path) -> bool {
+        self.r#match.is_match(canonical_path)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct AssetSourceDirectoryConfiguration {
+    config_map: ConfigMap,
+}
+
+#[derive(Debug, Default, PartialEq, Eq, Serialize, Clone)]
+pub(crate) struct AssetConfig {
+    pub(crate) cache: Option<CacheConfig>,
+    pub(crate) headers: Option<HeadersConfig>,
+}
+
+#[derive(Debug, Default)]
+struct AssetConfigTreeNode {
+    pub parent: Option<Arc<AssetConfigTreeNode>>,
+    pub rules: Vec<AssetConfigRule>,
+}
+
+impl AssetSourceDirectoryConfiguration {
+    pub(crate) fn load(root_dir: &Path) -> anyhow::Result<Self> {
+        let mut config_map = HashMap::new();
+        let root_dir = root_dir.canonicalize()?;
+        AssetConfigTreeNode::load(None, &root_dir, &mut config_map)?;
+
+        Ok(Self { config_map })
+    }
+
+    pub(crate) fn get_asset_config(&self, canonical_path: &Path) -> anyhow::Result<AssetConfig> {
+        let parent_dir = canonical_path.parent().with_context(|| {
+            format!(
+                "unable to get the parent directory for asset path: {:?}",
+                canonical_path
+            )
+        })?;
+        Ok(self
+            .config_map
+            .get(parent_dir)
+            .with_context(|| {
+                format!(
+                    "unable to find default config for following path: {:?}",
+                    parent_dir
+                )
+            })?
+            .get_config(canonical_path))
+    }
+}
+
+#[derive(Deserialize)]
+struct InterimAssetConfigRule {
+    r#match: String,
+    cache: Option<CacheConfig>,
+    #[serde(default, deserialize_with = "deser_headers")]
+    headers: Maybe<HeadersConfig>,
+}
+
+impl<T> Default for Maybe<T> {
+    fn default() -> Self {
+        Self::Absent
+    }
+}
+
+fn deser_headers<'de, D>(deserializer: D) -> Result<Maybe<HeadersConfig>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    match serde_json::value::Value::deserialize(deserializer)? {
+        Value::Object(v) => Ok(Maybe::Value(
+            v.into_iter().collect::<HashMap<String, Value>>(),
+        )),
+        Value::Null => Ok(Maybe::Null),
+        _ => Err(serde::de::Error::custom(
+            "wrong data format for field `headers` (only map or null are allowed)",
+        )),
+    }
+}
+
+impl AssetConfigRule {
+    fn from_interim(
+        InterimAssetConfigRule {
+            r#match,
+            cache,
+            headers,
+        }: InterimAssetConfigRule,
+        config_file_parent_dir: &Path,
+    ) -> anyhow::Result<Self> {
+        let glob = Glob::new(
+            config_file_parent_dir
+                .join(&r#match)
+                .to_str()
+                .with_context(|| format!("{} is not a valid glob pattern", r#match))?,
+        )
+        .with_context(|| format!("{} is not a valid glob pattern", r#match))?
```

Comment:
Can we use different error strings to make it easier to know which case failed?

Reply:
done in https://github.com/dfinity/agent-rs/pull/352/commits/49230c039dfcf71c614f94cfb304537d3145a630